step-functions/step-functions-client/src/main/java/step/functions/execution/FunctionExecutionServiceImpl.java

```diff
@@ -102,27 +102,69 @@ public FunctionExecutionServiceImpl(GridClient gridClient, FunctionTypeRegistry
 	}
 
 	private static final Logger logger = LoggerFactory.getLogger(FunctionExecutionServiceImpl.class);
-	
+
+	private final List<TokenLifecycleInterceptor> tokenLifecycleInterceptors = new ArrayList<>();
+
+	@Override
+	public void registerTokenLifecycleInterceptor(TokenLifecycleInterceptor interceptor) {
+		synchronized (tokenLifecycleInterceptors) {
+			tokenLifecycleInterceptors.add(interceptor);
+		}
+	}
+
+
+	@Override
+	public void unregisterTokenLifecycleInterceptor(TokenLifecycleInterceptor interceptor) {
+		synchronized (tokenLifecycleInterceptors) {
+			tokenLifecycleInterceptors.remove(interceptor);
+		}
+	}
+
 	@Override
 	public TokenWrapper getLocalTokenHandle() {
 		return gridClient.getLocalTokenHandle();
 	}
 
 	@Override
 	public TokenWrapper getTokenHandle(Map<String, String> attributes, Map<String, Interest> interests, boolean createSession, TokenWrapperOwner tokenWrapperOwner) throws FunctionExecutionServiceException {
+		TokenWrapper tokenWrapper = null;
 		try {
-			return gridClient.getTokenHandle(attributes, interests, createSession, tokenWrapperOwner);
+			tokenWrapper = gridClient.getTokenHandle(attributes, interests, createSession, tokenWrapperOwner);
 		} catch (AgentCallTimeoutException e) {
 			throw new FunctionExecutionServiceException("Timeout after "+e.getCallTimeout()+"ms while reserving the agent token. You can increase the call timeout by setting 'grid.client.token.reserve.timeout.ms' in step.properties",e );
 		} catch (AgentSideException e) {
 			throw new FunctionExecutionServiceException("Unexpected error on the agent side while reserving the agent token: "+e.getMessage(),e);
 		} catch (AgentCommunicationException e) {
 			throw new FunctionExecutionServiceException("Communication error between the controller and the agent while reserving the agent token",e);
-		} 
+		}
+
+		try {
+			for (TokenLifecycleInterceptor interceptor : getTokenLifecycleInterceptors()) {

```

Comment:
Use forEach instead

Reply:
Does not work, because catching an Exception inside foreach is not possible.